flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorExecutionDeploymentReconciliationTest.java

```diff
@@ -0,0 +1,324 @@
+			.setTaskManagerHeartbeatConsumer((resourceID, taskExecutorToJobManagerHeartbeatPayload) -> {
+				ExecutionDeploymentReport executionDeploymentReport = taskExecutorToJobManagerHeartbeatPayload.getExecutionDeploymentReport();
+				deployedExecutionsFuture.add(executionDeploymentReport.getExecutions());
+			})
+			.setUpdateTaskExecutionStateFunction(taskExecutionState -> {
+				if (taskExecutionState.getExecutionState() == ExecutionState.FINISHED) {
+					taskFinishedFuture.complete(null);
+				}
+				return CompletableFuture.completedFuture(Acknowledge.get());
+			})
+			.build();
+	}
+
+	private static TestingResourceManagerGateway setupResourceManagerGateway(CompletableFuture<SlotReport> initialSlotReportFuture) {
+		final TestingResourceManagerGateway testingResourceManagerGateway = new TestingResourceManagerGateway();
+		testingResourceManagerGateway.setSendSlotReportFunction(resourceIDInstanceIDSlotReportTuple3 -> {
+			initialSlotReportFuture.complete(resourceIDInstanceIDSlotReportTuple3.f2);
+			return CompletableFuture.completedFuture(Acknowledge.get());
+		});
+		testingResourceManagerGateway.setRegisterTaskExecutorFunction(input -> CompletableFuture.completedFuture(
+			new TaskExecutorRegistrationSuccess(
+				new InstanceID(),
+				testingResourceManagerGateway.getOwnResourceId(),
+				new ClusterInformation("blobServerHost", 55555))));
+		return testingResourceManagerGateway;
+	}
+
+	private void connectComponentsAndRequestSlot(
+			JobMasterGateway jobMasterGateway,
+			ResourceManagerGateway resourceManagerGateway,
+			TaskExecutorGateway taskExecutorGateway,
+			JobLeaderService jobLeaderService,
+			CompletableFuture<SlotReport> initialSlotReportFuture,
+			AllocationID allocationId) throws Exception {
+		final String jobMasterAddress = "jm";
+		rpc.registerGateway(jobMasterAddress, jobMasterGateway);
+		rpc.registerGateway(resourceManagerGateway.getAddress(), resourceManagerGateway);
+
+		// inform the task manager about the job leader
+		jobLeaderService.addJob(jobId, jobMasterAddress);
+		jobManagerLeaderRetriever.notifyListener(jobMasterAddress, UUID.randomUUID());
+		resourceManagerLeaderRetriever.notifyListener(resourceManagerGateway.getAddress(), resourceManagerGateway.getFencingToken().toUUID());
+
+		final Optional<SlotStatus> slotStatusOptional = StreamSupport.stream(initialSlotReportFuture.get().spliterator(), false)
+			.findAny();
+
+		assertTrue(slotStatusOptional.isPresent());
+
+		while (true) {
+			try {
+				taskExecutorGateway.requestSlot(
+					slotStatusOptional.get().getSlotID(),
+					jobId,
+					allocationId,
+					ResourceProfile.ZERO,
+					jobMasterAddress,
+					resourceManagerGateway.getFencingToken(),
+					timeout
+				).get();
+				break;
+			} catch (Exception e) {
+				// the proper establishment of the RM connection is tracked
+				// asynchronously, so we have to poll here until it went through
+				// until then, slot requests will fail with an exception
+				Thread.sleep(50);
+			}
+		}
```

Comment:
If you want to wait that the `TaskExecutor` has established the connection to the `ResourceManager`, then one can wait for the `ResourceManagerGateway.sendSlotReport` call.